Optimize memory usage when merging genotypes

VJalili opened the pull request:
The script `merge_RdTest_genotypes.py` runs out of memory when executed on relatively small files (files of size `1.0GB` and `600MB` for arguments `genotypes` and `GQ` respectively). This PR optimizes this script.

## First try ([fc8ba1b](https://github.com/broadinstitute/gatk-sv/pull/175/commits/fc8ba1b597170ea0ddaad128241e6b9346e0e8b5)): Improve on object types and replace `melt` with `stack`

- The genotype of samples is a relatively small number, however, there are few samples with values set to `.` which causes all the values to be parsed as an `object` type. Accordingly, any rows with values `.` are removed (using a flag).
- The values are read as `uint8` (maybe we need to change it to `uint16`) which helps reduce the memory usage. 
- Panda's `melt` does not use categorical representations, hence it increases object size a fewfold. This PR replaces `melt` with `stack` (whose outputs are comparable, only ordering may differ). 

### Data types 

```
# As read from the input file.
>>> print(gt.dtypes)
chr                           object
start                          int64
end                            int64
cnvID                         object
__fes_0000_0__000000           uint8
                               ...  
__tps_0000000_01_a__000000     uint8
__tps_0000000_02_a__000000     uint8
__tps_00000000__000000         uint8
__tps_0000000_03_a__000000     uint8
__tps_0000000_04_a__000000     uint8
Length: 334, dtype: object

# After stack.
>>> print(gt.dtypes)
chr           object
start          int64
end            int64
cnvID         object
sample      category
genotype       uint8
dtype: object
```

### Performance

| Object        | Size (master)        | Size (commit [fc8ba1b](https://github.com/broadinstitute/gatk-sv/commit/fc8ba1b597170ea0ddaad128241e6b9346e0e8b5))        |
|---------------|----------------------|-----------------------|
| `gt` from input | 9.84 GB | 0.41 GB  |
| `gt` stacked    | 75.86 GB | 46.51 GB  |
| `gq` from input | 9.81 GB  | 0.41 GB |
| `gq` stacked    | 75.82 GB | 46.78 GB |
| merged        | failed               | 49.43 GB  |

## Second try ([7464206](https://github.com/broadinstitute/gatk-sv/pull/175/commits/74642061ebfddf5d541b2176b8465baf99979f60)): Reimplement the script without loading the data in memory

[7464206](https://github.com/broadinstitute/gatk-sv/pull/175/commits/74642061ebfddf5d541b2176b8465baf99979f60) reads files line-by-line---reading genotypes and GQ files simultaneously---and melts/merges them on the fly, and writes them to a file as it proceeds. Hence, it minimizes memory usage to a few small objects. 

The script processed genotype file of size `568MB` and GQ of size `1.0GB` in `2min 40sec`:

```
$ ls -lha
-rw-r--r--@  1 vahid  staff   1.0G May 13 15:23 rd.GQ.all
-rw-r--r--@  1 vahid  staff   596M May 13 15:17 rd.geno.all

$ time python merge_RdTest_genotypes.py rd.geno.all rd.GQ.all merged.bed

real	2m40.295s
user	2m7.529s
sys	0m29.137s
```

mwalker174 reviewed (commented):
This is excellent! Thanks for providing all the benchmarks as well (I assume the output were all byte-identical in your tests). 

Just need to update the `sv-pipeline` docker. I've sent you a DM on how to do that.

VJalili commented:
> I assume the output were all byte-identical in your tests

VJalili commented:
> I assume the output were all byte-identical in your tests

Thank you @mwalker174 for reviewing this. The two files seem identical, other than their ordering. Neither of the files is sorted, hence I guess the ordering should not matter down the line. For instance: 

```shell
$ head merge_master_branch.bed 
chr1	667949	668041	batch1_manta_chr1_chr1_18	HG00096	1	116
chr1	820394	823548	batch1_wham_chr1_chr1_6	HG00096	2	999
chr1	820717	821270	batch1_wham_chr1_chr1_7	HG00096	3	132
chr1	822590	823465	batch1_wham_chr1_chr1_8	HG00096	2	43
chr1	830860	833732	batch1_wham_chr1_chr1_10	HG00096	2	121
chr1	831215	833732	batch1_manta_chr1_chr1_29	HG00096	2	121
chr1	904468	904739	batch1_wham_chr1_chr1_12	HG00096	2	149
chr1	906945	907045	batch1_manta_chr1_chr1_37	HG00096	3	61
chr1	917444	917517	batch1_manta_chr1_chr1_38	HG00096	2	65
chr1	917444	917517	batch1_wham_chr1_chr1_14	HG00096	2	65
$ head merge_feature_branch.bed 
chr1	667949	668041	batch1_manta_chr1_chr1_18	HG00096	1	116
chr1	667949	668041	batch1_manta_chr1_chr1_18	HG00129	2	12
chr1	667949	668041	batch1_manta_chr1_chr1_18	HG00140	2	94
chr1	667949	668041	batch1_manta_chr1_chr1_18	HG00150	1	134
chr1	667949	668041	batch1_manta_chr1_chr1_18	HG00187	2	23
chr1	667949	668041	batch1_manta_chr1_chr1_18	HG00239	0	132
chr1	667949	668041	batch1_manta_chr1_chr1_18	HG00277	1	128
chr1	667949	668041	batch1_manta_chr1_chr1_18	HG00288	1	128
chr1	667949	668041	batch1_manta_chr1_chr1_18	HG00337	2	18
chr1	667949	668041	batch1_manta_chr1_chr1_18	HG00349	2	73
```

I compared the outputs as the following: 

```shell
# Run using the script on the master branch
$ git checkout master && git rev-parse HEAD
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
22fc647dd3ee56a47cb0523ea8f6cc37c4b904eb
$ time python merge_RdTest_genotypes.py rd.geno.all rd.GQ.all merge_master_branch.bed

real	0m40.914s
user	0m37.206s
sys	0m2.964s

# Run using the script on this branch 
$ git checkout merge_genotypes && git rev-parse HEAD
Switched to branch 'merge_genotypes'
Your branch is up to date with 'origin/merge_genotypes'.
17e26ff10d4ec4aa694dc378a7bfaa9da7e9d3b2
$ time python merge_RdTest_genotypes.py rd.geno.all rd.GQ.all merge_feature_branch.bed

real	0m4.653s
user	0m3.939s
sys	0m0.608s

# Compare outputs
$ sort -o merge_feature_branch_sorted.bed merge_feature_branch.bed && sort -o merge_master_branch_sorted.bed merge_master_branch.bed
$ diff merge_master_branch_sorted.bed merge_feature_branch_sorted.bed 
$ ls -l merge_*_branch_sorted.bed
-rw-r--r--  1 vahid  staff  580465907 Jun  7 13:44 merge_feature_branch_sorted.bed
-rw-r--r--  1 vahid  staff  580465907 Jun  7 13:45 merge_master_branch_sorted.bed
```

VJalili commented:
@mwalker174 To make sure the docker images build and run successfully, with the help of @epiercehoffman, I ran module 04 and it finished successfully. 

```shell
Sub-Command: status
Using workflow-id == bacaa818-cfe6-4d2f-b11d-62e43f417fa8
Using workflow server URL == https://cromwell-v54.dsde-methods.broadinstitute.org/
                  __     
       .,-;-;-,. /'_\    
     _/_/_/_|_\_\) /     
   '-<_><_><_><_>=/\     
     `/_/====/_/-'\_\    
      ""     ""    ""    
{
  "status": "Succeeded",
  "id": "bacaa818-cfe6-4d2f-b11d-62e43f417fa8"
}
```

I am not sure if `"Succeeded"` is good enough, but happy to check the output files if necessary.

epiercehoffman commented:
> I am not sure if `"Succeeded"` is good enough, but happy to check the output files if necessary.

Since you verified that the output of the script is identical after sorting, I think you are probably fine with just a successful run of the module. But if you want to be extra careful, you could compare with the latest master and verify the outputs are identical (just the VCFs is fine I think).

After that, make sure you update `dockers.json` with your new docker images.

mwalker174 commented:
Check that the module04 outputs are nearly the same as those in the Terra workspace.

VJalili commented:
The input json I used is:

```json
{
  "Module04.sv_base_mini_docker":"us.gcr.io/broad-dsde-methods/vjalili/sv-base-mini:merge-genotypes-fc8ba1b",
  "Module04.sv_pipeline_docker": "us.gcr.io/broad-dsde-methods/vjalili/sv-pipeline:merge-genotypes-fc8ba1b",
  "Module04.sv_pipeline_rdtest_docker": "us.gcr.io/broad-dsde-methods/cwhelan/sv-pipeline-rdtest:cw_rdtest_bin_fix_update_5f2c8f",
  "Module04.linux_docker" : "marketplace.gcr.io/google/ubuntu1804",

  "Module04.n_RD_genotype_bins": "100000", 
  "Module04.n_per_split": "5000", 
  "Module04.pesr_exclude_list": "gs://gcp-public-data--broad-references/hg38/v0/sv-resources/resources/v1/PESR.encode.peri_all.repeats.delly.hg38.blacklist.bed.gz",
  "Module04.seed_cutoffs": "gs://gatk-sv-resources-public/hg38/v0/sv-resources/resources/v1/seed_cutoff.txt",
  "Module04.reference_build": "hg38",
  "Module04.ref_dict": "gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dict",

  "Module04.batch": "test_large",
  "Module04.rf_cutoffs": "gs://gatk-sv-resources/test/module03/large/output/test_large.cutoffs",
  "Module04.batch_depth_vcf": "gs://gatk-sv-resources/test/module03/large/output/test_large.depth.outliers_removed.vcf.gz",
  "Module04.batch_pesr_vcf": "gs://gatk-sv-resources/test/module03/large/output/test_large.filtered_pesr_merged.vcf.gz",
  "Module04.ped_file": "gs://gatk-sv-resources/test/resources/ref_panel_v1.ped",
  "Module04.bin_exclude": "gs://gatk-sv-resources-public/hg38/v0/sv-resources/resources/v1/bin_exclude.hg38.gatkcov.bed.gz",
  "Module04.discfile": "gs://gatk-sv-resources/test/module00c/large/outputs/test_large.PE.txt.gz",
  "Module04.coveragefile": "gs://gatk-sv-resources/test/module00c/large/outputs/test_large.RD.txt.gz",
  "Module04.splitfile": "gs://gatk-sv-resources/test/module00c/large/outputs/test_large.SR.txt.gz",
  "Module04.medianfile": "gs://gatk-sv-resources/test/module00c/large/outputs/test_large_medianCov.transposed.bed",
  "Module04.cohort_depth_vcf": "gs://talkowskicromwellresults/M04/MergeCohortVcfs/190da1da-75b8-485e-82fc-b2c5ff07a8e6/call-MergeDepthVcfs/all_batches.depth.vcf.gz",
  "Module04.cohort_pesr_vcf": "gs://talkowskicromwellresults/M04/MergeCohortVcfs/190da1da-75b8-485e-82fc-b2c5ff07a8e6/call-MergePESRVcfs/all_batches.pesr.vcf.gz"
}
```

VJalili commented:
@epiercehoffman Agree, that will be a better assurance. I ran the pipeline using the current latest master and this branch, and it seems their VCF files are bitwise identical (see below); if I am comparing the correct VCF files.

- This branch:
    ```
    $ cromshell status b6c9bec5-f79a-4d97-9d27-a2394a2fac04
    Sub-Command: status
    Using workflow-id == b6c9bec5-f79a-4d97-9d27-a2394a2fac04
    Using workflow server URL == https://cromwell-talkowski.dsde-methods.broadinstitute.org/
                      __     
           .,-;-;-,. /'_\    
         _/_/_/_|_\_\) /     
       '-<_><_><_><_>=/\     
         `/_/====/_/-'\_\    
          ""     ""    ""    
    {
      "status": "Succeeded",
      "id": "b6c9bec5-f79a-4d97-9d27-a2394a2fac04"
    }

    $ gsutil cp gs://talkowski-sv-gnomad-cromwell-execution/Module04/b6c9bec5-f79a-4d97-9d27-a2394a2fac04/call-GenotypePESRPart2/GenotypePESRPart2/e0a38925-96e1-4f92-ae46-f056c8bc6328/call-ConcatGenotypedVcfs/test_large.pesr.vcf.gz .
    $ gsutil cp gs://talkowski-sv-gnomad-cromwell-execution/Module04/b6c9bec5-f79a-4d97-9d27-a2394a2fac04/call-GenotypePESRPart2/GenotypePESRPart2/e0a38925-96e1-4f92-ae46-f056c8bc6328/call-ConcatGenotypedVcfs/test_large.pesr.vcf.gz.tbi .
    $ gsutil cp gs://talkowski-sv-gnomad-cromwell-execution/Module04/b6c9bec5-f79a-4d97-9d27-a2394a2fac04/call-GenotypeDepthPart2/GenotypeDepthPart2/278db40b-99cc-473d-b7d4-f3202f5a2e09/call-ConcatGenotypedVcfs/test_large.depth.vcf.gz .
    $ gunzip test_large.depth.vcf.gz
    $ gunzip csv/test_large.pesr.vcf.gz

    $ ls -l
    total 336648
    -rw-r--r--  1 vahid  staff   14245309 Jun 14 22:00 test_large.depth.vcf
    -rw-r--r--  1 vahid  staff  157855221 Jun 14 21:59 test_large.pesr.vcf
    -rw-r--r--  1 vahid  staff     245853 Jun 14 21:59 test_large.pesr.vcf.gz.tbi
    ```

- Current Latest master branch:
    ```
    $ cromshell status 34a3f7d8-9cf4-4322-8df3-64803c209d19
    Sub-Command: status
    Using workflow-id == 34a3f7d8-9cf4-4322-8df3-64803c209d19
    Using workflow server URL == https://cromwell-v54.dsde-methods.broadinstitute.org/
                      __     
           .,-;-;-,. /'_\    
         _/_/_/_|_\_\) /     
       '-<_><_><_><_>=/\     
         `/_/====/_/-'\_\    
          ""     ""    ""    
    {
      "status": "Succeeded",
      "id": "34a3f7d8-9cf4-4322-8df3-64803c209d19"
    }

    $ gsutil cp gs://broad-methods-cromwell-exec-bucket-v51/Module04/34a3f7d8-9cf4-4322-8df3-64803c209d19/call-GenotypePESRPart2/GenotypePESRPart2/ad4b6a98-6319-4007-acd9-b040d5554334/call-ConcatGenotypedVcfs/test_large.pesr.vcf.gz .
    $ gsutil cp gs://broad-methods-cromwell-exec-bucket-v51/Module04/34a3f7d8-9cf4-4322-8df3-64803c209d19/call-GenotypePESRPart2/GenotypePESRPart2/ad4b6a98-6319-4007-acd9-b040d5554334/call-ConcatGenotypedVcfs/test_large.pesr.vcf.gz.tbi .
    $ gsutil cp gs://broad-methods-cromwell-exec-bucket-v51/Module04/34a3f7d8-9cf4-4322-8df3-64803c209d19/call-GenotypeDepthPart2/GenotypeDepthPart2/86790481-e320-4d45-a2e4-f3e5e09377a1/call-ConcatGenotypedVcfs/test_large.depth.vcf.gz .

    $ ls -l
    total 336744
    -rw-r--r--  1 vahid  staff   14245309 Jun 14 22:03 test_large.depth.vcf
    -rw-r--r--  1 vahid  staff  157855221 Jun 14 22:03 test_large.pesr.vcf
    -rw-r--r--  1 vahid  staff     245853 Jun 14 22:03 test_large.pesr.vcf.gz.tbi
    ```

VJalili commented on input_values/dockers.json:
```diff
@@ -12,8 +12,8 @@
   "sv_base_docker" : "us.gcr.io/broad-dsde-methods/gatk-sv/sv-base:mw-gnomad-02-6a66c96",
   "sv_base_mini_docker" : "us.gcr.io/broad-dsde-methods/gatk-sv/sv-base-mini:rlc_posthoc_filtering_cnv_mcnv_compatability_9a8561",
-  "sv_pipeline_base_docker" : "us.gcr.io/broad-dsde-methods/gatk-sv/sv-pipeline-base:rlc_posthoc_filtering_cnv_mcnv_compatability_9a8561",
-  "sv_pipeline_docker" : "us.gcr.io/broad-dsde-methods/markw/sv-pipeline:mw-clean-vcf-4-sr-gq-bd75b73",
```
@mwalker174 are these the only two images that should be updated?

mwalker174 replied:
Yes
